draft-jholland-quic-multicast.md

```diff
@@ -191,6 +191,8 @@ The client tells its server about some restrictions on resources that it is capa
 The server asks the client to join channels with MC_JOIN ({{channel-join-frame}}) frames and to leave channels with MC_LEAVE ({{channel-leave-frame}}) frames.
 
 The server uses the MC_ANNOUNCE ({{channel-announce-frame}}) frame before any join or leave frames for the channel to describe the channel properties to the client, including values the client can use to ensure the server's requests remain within the limits it has sent to the server, as well as the secrets necessary to decode the headers of packets in the channel.
+Sending an MC_ANNOUNCE before an MC_JOIN ensures the client can establish the necessary state required to join and retire any connection IDs that might collide with channel IDs.
+Separating announcement of channels from joining them allows clients to leave and later rejoining channels at their discretion without the cost of terminating a connection. This could, for example, be used by clients to temporarily disabling multicast reception in response to a high rate of loss or a switch to a network that does not support multicast.
```

Comment:
this sentence has a few problems I think:

- nits: "leave and later rejoining channels" -> "leave and later rejoin channels", and similarly: "used by clients to temporarily disabling" -> "used by clients to temporarily disable"
- "at their discretion" applies only to the leaving, they cannot rejoin if they've sent a MC_STATE(Leave) unless server instructs them to do so.  Unless you're saying they could leave at the network layer without the corresponding state change in the quic connection telling the server they left?  I was assuming that's not allowed and would indicate a buggy client, but maybe we should explicitly state our expectations about that somewhere...

Reply:
agree with all points

Reply:
Huh, I had more typed in here, not sure how it got lost.

I continued to say that maybe we should drop the 2nd line justifying why announce is separated from join, and discuss that offline instead.

I don't think the doc has to include justification for all the design decisions, and we should maybe discuss offline whether and why we think it's the right design decision, as I don't think this quite covers it to me.  (I don't expect it would cost terminating a connection to leave and rejoin, for instance.)

I do think it's right to keep join and announce separated, but I think the biggest reason is about being able to assume stability of the channel properties during a join/leave cycle, since I think you get more weird corner cases if client has to handle a leave followed by a join that has different properties, especially once you think about what happens if a frame was dropped.

I think separating channel existence into announce/retire and channel subscription into join/leave probably makes the implementations tend to be cleaner, assuming we can legitimately get somewhat frequent join/leave cycles on the order of a few seconds, but I think we can just say that in email instead of in the doc.

Reply:
(the 2nd biggest reason is that it's something like 20x bigger to do a leave+re-join, which seems like a lot to me even for control junk, since I do think we expect it in normal operation sometimes.)

Reply:
> I think separating channel existence into announce/retire and channel subscription into join/leave probably makes the implementations tend to be cleaner, assuming we can legitimately get somewhat frequent join/leave cycles on the order of a few seconds, but I think we can just say that in email instead of in the doc.

Yeah I tend to agree. There's techniques that application can employ to make channel hopping beneficial, especially when the delay is minimal. The transport specification doesn't need to get bogged down with that detail, if we agree elsewhere that is something beneficial to support.

Reply:
Sure, I agree that maybe the justification doesn't have to be in the draft, removed the line and noted it down for later justification of design decisions if necessary (I made an issue in my fork, I will add any similar things if they pop up there so we can keep track, feel free to add any additional ones as well: https://github.com/MaxF12/draft-jholland-quic-multicast/issues/4)